File: scode/KoalaCoders/prob13/main.c
#include <stdio.h>

int pgcd(int x, int y) {
    while (y != 0) {
        int temp = y;
        y = x % y;
        x = temp;
    }
    return x;
}

int pgcdl(int l[], int size) {
    if (size == 1) {
        return pgcd(l[0], l[0]);
    }
    if (pgcd(l[0], l[1]) == 1) {
        return 1;
    }
    l[0] = pgcd(l[0], l[1]);
    for (int i = 1; i < size - 1; i++) {
        l[i] = l[i + 1];
    }
    return pgcdl(l, size - 1);
}

void generate(int l[], int size, int comb[][10], int* comb_size) {
    *comb_size = 0;
    comb[*comb_size][0] = 0;
    (*comb_size)++;
    for (int i = 0; i < size; i++) {
        int temp_size = *comb_size;
        for (int j = 0; j < temp_size; j++) {
            for (int k = 0; k < i + 1; k++) {
                comb[*comb_size][k] = comb[j][k];
            }
            comb[*comb_size][i + 1] = l[i];
            (*comb_size)++;
        }
    }
}

int main() {
    int n;
    scanf("%d", &n);
    int h[100];
    for (int i = 0; i < n; i++) {
        scanf("%d", &h[i]);
    }
    int d[100][10];
    int d_size[100] = {0};
    int s = 0;
    generate(h, n, d, &d_size[0]);
    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < d_size[i]; j++) {
            if (pgcdl(d[i], i + 1) != 1) {
                s += pgcdl(d[i], i + 1) * (i + 1);
            }
        }
    }
    printf("%d", s);
    return 0;
}

File: scode/KoalaCoders/prob13/main.py


def pgcd(x, y):
    while y != 0:
        x, y = y, x % y
    return x

def pgcdl(l):
    if len(l)==1:
        return pgcd(l[0],l[0])
    if pgcd(l[0],l[1])==1 or 1 in l:
        return 1
    l[0]=pgcd(l[0],l[1])
    l.remove(l[1])
    return pgcdl(l)

def generate(l):
    comb = [[]]
    for e in l:
        comb += [sub + [e] for sub in comb]
    return comb[1:]


n=int(input())
h = list(map(int, input().split()))
d={}
for k in generate(h):
    long=len(k)
    if long in d:
        d[long].append(k)
    else:
        d[long]=[k]

s=0
for i in d:
    for j in range(len(d[i])):
        if pgcdl(d[i][j])!=1:
            s+=pgcdl(d[i][j])*i
print(s)




     